include/miroil/miroil/event_builder.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (C) 2016-2021 Canonical, Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3, as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+ * SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIROIL_EVENT_BUILDER_H
+#define MIROIL_EVENT_BUILDER_H
+#include <miral/toolkit_event.h>
+#include <mir_toolkit/mir_input_device_types.h>
+#include <chrono>
+#include <memory>
+#include <sys/types.h>
+#include <vector>
+
+struct MirEvent;
+struct MirInputEvent;
+
+namespace mir { typedef std::unique_ptr<MirEvent, void(*)(MirEvent*)> EventUPtr; }
+
+namespace miroil
+{
+    
+/*
+    Creates Mir input events out of Qt input events
+    
+    The class is splitt into miroil::EventBuilder which does the internal mir stuff, 
+    and qtmir::EventBuilder which handles the qt stuff. 
+
+    One important feature is that it's able to match a QInputEvent with the MirInputEvent that originated it, so
+    it can make a MirInputEvent version of a QInputEvent containing also information that the latter does not carry,
+    such as relative axis movement for pointer devices.
+*/
+    
+class EventBuilder {
+    
+public:
+    class EventInfo {
+    public:
+        void store(const MirInputEvent *mirInputEvent, ulong qtTimestamp);
+        
+        ulong timestamp;
+        MirInputDeviceId device_id;
+        std::vector<uint8_t> cookie;
+        float relative_x{0};
+        float relative_y{0};
+    };
+    
+public:
+    EventBuilder();
+    virtual ~EventBuilder() {};    
+    
+    // add Touch event
+    void add_touch(MirEvent &event, MirTouchId touch_id, MirTouchAction action,
+        MirTouchTooltype tooltype, float x_axis_value, float y_axis_value,
+        float pressure_value, float touch_major_value, float touch_minor_value, float size_value);
+    
+    // Key event
+    mir::EventUPtr make_event(MirInputDeviceId device_id, std::chrono::nanoseconds timestamp,
+        std::vector<uint8_t> const& cookie, MirKeyboardAction action, xkb_keysym_t key_code,
+        int scan_code, MirInputEventModifiers modifiers);
+
+    // Touch event
+    mir::EventUPtr make_event(MirInputDeviceId device_id, std::chrono::nanoseconds timestamp,
+        std::vector<uint8_t> const& mac, MirInputEventModifiers modifiers);
```

Comment:
I know the mirclient API has this horrible abuse of overloading `make_event()`, but I don't think it right to repeat the mistake in miroil. Would it be painful in the QtMir code to have `make_touch_event()`, `make_key_event()`, etc?

Reply:
Yes, that sound like a good idea and is quick to do. So I'll change the names.

Reply:
Fixed that, and talk to Marius Gripsgard about branches... So https://github.com/ubports/qtmir/tree/xenial_-_edge_-_wayland_-_x11it the correct one to use for now.